File: problems/图/417.太平洋大西洋水流问题.cpp
#include <iostream>
#include <vector>
#include <utility>
using namespace std;

vector<vector<int>> pacificAtlantic(vector<vector<int>>& h) {
    int m = h.size(), n = h[0].size();
    vector<vector<char>> pac(m, vector<char>(n, 0)), atl(m, vector<char>(n, 0));
    function<void(int,int,vector<vector<char>>&,int)> dfs = [&](int r, int c, vector<vector<char>>& v, int h0) {
        if (r < 0 || r >= m || c < 0 || c >= n || v[r][c] || h[r][c] < h0) return;
        v[r][c] = 1;
        for (auto& d : vector<pair<int,int>>{{0,1},{0,-1},{1,0},{-1,0}})
            dfs(r+d.first, c+d.second, v, h[r][c]);
    };
    for (int c = 0; c < n; c++) dfs(0, c, pac, 0);
    for (int r = 0; r < m; r++) dfs(r, 0, pac, 0);
    for (int c = 0; c < n; c++) dfs(m-1, c, atl, 0);
    for (int r = 0; r < m; r++) dfs(r, n-1, atl, 0);
    vector<vector<int>> result;
    for (int r = 0; r < m; r++)
        for (int c = 0; c < n; c++)
            if (pac[r][c] && atl[r][c])
                result.push_back({r, c});
    return result;
}

int main() {
    // 测试用例
    vector<vector<int>> h1 = {{1,2,2,3,5},{3,2,3,4,4},{2,4,5,3,1},{6,7,1,4,5},{5,1,1,2,4}};
    auto r1 = pacificAtlantic(h1);
    for (auto& p : r1) cout << "(" << p[0] << "," << p[1] << ") "; cout << endl;
    // 预期: (0,4)(1,3)(1,4)(2,2)(3,0)(3,1)(4,0)
    vector<vector<int>> h2 = {{1}};
    auto r2 = pacificAtlantic(h2);
    for (auto& p : r2) cout << "(" << p[0] << "," << p[1] << ") "; cout << endl;  // 预期: (0,0)
    return 0;
}

File: problems/图/417.太平洋大西洋水流问题.py
#!/usr/bin/env python3
"""
417. 太平洋大西洋水流问题
"""
def pacific_atlantic(heights: list[list[int]]) -> list[list[int]]:
    if not heights:
        return []
    m, n = len(heights), len(heights[0])
    pac = [[False] * n for _ in range(m)]
    atl = [[False] * n for _ in range(m)]

    def dfs(r, c, visited, prev_h):
        if r < 0 or r >= m or c < 0 or c >= n or visited[r][c] or heights[r][c] < prev_h:
            return
        visited[r][c] = True
        for dr, dc in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
            dfs(r+dr, c+dc, visited, heights[r][c])

    for c in range(n): dfs(0, c, pac, 0)
    for r in range(m): dfs(r, 0, pac, 0)
    for c in range(n): dfs(m-1, c, atl, 0)
    for r in range(m): dfs(r, n-1, atl, 0)

    return [[r, c] for r in range(m) for c in range(n) if pac[r][c] and atl[r][c]]

if __name__ == "__main__":
    # 测试用例
    heights1 = [[1, 2, 2, 3, 5], [3, 2, 3, 4, 4], [2, 4, 5, 3, 1], [6, 7, 1, 4, 5], [5, 1, 1, 2, 4]]
    result1 = pacific_atlantic(heights1)
    print(result1)  # 预期: [[0,4],[1,3],[1,4],[2,2],[3,0],[3,1],[4,0]]
    heights2 = [[1]]
    result2 = pacific_atlantic(heights2)
    print(result2)  # 预期: [[0,0]]
